### PR TITLE
feat(consensus): add central invoke transaction

### PR DESCRIPTION
**Pull request by Yael-Starkware**

### CODE REVIEW

**Comment by Yael-Starkware** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2861?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2861** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2861?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2861?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2715** <a href="https://app.graphite.dev/github/pr/starkware-libs/sequencer/2715?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs [hunk 1]
```diff
@@ -12,9 +14,30 @@ use starknet_api::block::{
...
+    CentralBlockInfo,
+    CentralInvokeTransaction,
+    CentralInvokeTransactionV3,
+    CentralResourcePrice,
+    CentralStateDiff,
+    CentralTransaction,
+    CentralTransactionTimestamp,
+    CentralTransactionWritten,
+};
+>>>>>>> c4145022a (feat(consensus): add central invoke transaction)
```
[the remark this answers is not in the document]
Comment (Yael-Starkware) [reply crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs]: Done.

**Review by Yael-Starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: 0 of 5 files reviewed, 1 unresolved discussion (waiting on @dafnamatsry, @DvirYo-starkware, and @graphite-app[bot])

**Review by dafnamatsry — CHANGES_REQUESTED** [review]
> Reviewed 2 of 5 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 7 unresolved discussions (waiting on @DvirYo-starkware, @graphite-app[bot], and @Yael-Starkware)

___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 79 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErlN3B7Fkoz2o8uPS0:-OErlN3B7Fkoz2o8uPS1:b-682cmz) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L79)):*
> ```XML
> pub struct CentralStateDiff {
>     pub address_to_class_hash: IndexMap<ContractAddress, ClassHash>,
>     pub nonces: IndexMap<DataAvailabilityMode, IndexMap<ContractAddress, Nonce>>,
> ```

Replace with `CentralDataAvailabilityMode`?



_Code quote:_
```XML
DataAvailabilityMode
```
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 126 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErmVqZ5gWMoJg_kqei:-OErmVqZ5gWMoJg_kqej:b-ax86l5) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L126)):*
> ```XML
> 
> #[derive(Debug, PartialEq, Deserialize, Serialize)]
> pub struct CentralTransactionTimestamp(pub u64);
> ```

Do we need this extra type? why not use `u64` directly?
Anyway, please move to be above `CentralTransactionWritten` where it is used.



_Code quote:_
```XML
pub struct CentralTransactionTimestamp(pub u64);
```
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 134 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErtTiz-OgJ6LLHyHHD:-OErtTiz-OgJ6LLHyHHE:b-jhk7sc) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L134)):*
> ```XML
>     pub signature: TransactionSignature,
>     pub nonce: Nonce,
>     pub resource_bounds: ValidResourceBounds,
> ```

Should we have a central version for this too?



_Code quote:_
```XML
ValidResourceBounds
```
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 144 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErylYj02nQ9E2CA3XP:-OErylYj02nQ9E2CA3XQ:bis6niz) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L144)):*
> ```XML
> 
> impl From<InvokeTransaction> for CentralInvokeTransactionV3 {
>     fn from(tx: InvokeTransaction) -> CentralInvokeTransactionV3 {
> ```

Please enforce it to be  `InvokeTransactionV3`



_Code quote:_
```XML
InvokeTransaction
```
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs` line 75 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OEs0EKL7kb93yMGlbq2:-OEs0EKL7kb93yMGlbq3:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs#L75)):*
> ```XML
> }
> 
> fn invoke_transaction() -> CentralTransactionWritten {
> ```





_Suggestion:_
```XML
central_invoke_transaction()
```
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs` line 102 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OEs-M4pDnXmcXpgdNtg:-OEs-M4pDnXmcXpgdNth:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs#L102)):*
> ```XML
>         })),
>         time_created: CentralTransactionTimestamp(1734601615),
>     }
> ```





_Suggestion:_
```XML
    CentralTransactionWritten {
        tx: CentralTransaction::Invoke(CentralInvokeTransaction::V3(CentralInvokeTransactionV3 {
            resource_bounds: ValidResourceBounds::AllResources(AllResourceBounds {
                l1_gas: ResourceBounds {
                    max_amount: GasAmount(1),
                    max_price_per_unit: GasPrice(1),
                },
                l2_gas: ResourceBounds::default(),
                l1_data_gas: ResourceBounds::default(),
            }),
            sender_address: contract_address!(
                "0x14abfd58671a1a9b30de2fcd2a42e8bff2ce1096a7c70bc7995904965f277e"
            ),
            calldata: Calldata(Arc::new(vec![felt!(0_u8), felt!(1_u8)])),
            hash_value: TransactionHash(felt!(
                "0x6efd067c859e6469d0f6d158e9ae408a9552eb8cc11f618ab3aef3e52450666"
            )),
            ..Default::default()
        })),
        time_created: CentralTransactionTimestamp(1734601615),
    }
```

**Review by Yael-Starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 7 unresolved discussions (waiting on @dafnamatsry, @DvirYo-starkware, and @graphite-app[bot])

___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 79 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErlN3B7Fkoz2o8uPS0:-OEsavy045pB3Bg_Uqy0:b6rlcv7) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L79)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Replace with `CentralDataAvailabilityMode`?
</blockquote></details>

the python code is not consistent in this regard, 
here in the StateDiff they are using enum which serializes to a string ("L1" or "L2"), whereas in the transactions they use an integer, with a limited range 0-1.
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 126 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErmVqZ5gWMoJg_kqei:-OEscHKj30JtpuObIzxt:b-g53689) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L126)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Do we need this extra type? why not use `u64` directly?
Anyway, please move to be above `CentralTransactionWritten` where it is used.
</blockquote></details>

we don't need it for serialization, but I did want to differentiate it from the BlockTimestamp to prevent errors.
moved.
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 134 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErtTiz-OgJ6LLHyHHD:-OEscmdZ3g11OsP2yKVJ:b-408638) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L134)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Should we have a central version for this too?
</blockquote></details>

This object is being serialized well to the python object using the starknet-api serialization, so we can use it as is.
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 144 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErylYj02nQ9E2CA3XP:-OEseUja3QkqMgUdg4B9:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L144)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Please enforce it to be  `InvokeTransactionV3`
</blockquote></details>

Done.
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs` line 75 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OEs0EKL7kb93yMGlbq2:-OEsg2G0-Qgw1-C5jv0U:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs#L75)):*
> ```XML
> }
> 
> fn invoke_transaction() -> CentralTransactionWritten {
> ```

Done.
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs` line 102 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OEs-M4pDnXmcXpgdNtg:-OEsfS6P8xuK_2Le85R_:b-1du6ad) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs#L102)):*
> ```XML
>         })),
>         time_created: CentralTransactionTimestamp(1734601615),
>     }
> ```

unfortunately there is no automatic default implementation for this object since it includes enum variants with data.

**Review by dafnamatsry — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 4 unresolved discussions (waiting on @DvirYo-starkware, @graphite-app[bot], and @Yael-Starkware)

___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 134 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErtTiz-OgJ6LLHyHHD:-OEwWyjl5QxFS31kFtPu:b-kuvrzb) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L134)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

This object is being serialized well to the python object using the starknet-api serialization, so we can use it as is.
</blockquote></details>

Might be a bit risky to rely on complex objects that we are not managing a can potentially change.
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 143 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OEwXh428AY5Losj4mW4:-OEwXh428AY5Losj4mW5:b7cd5hq) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L143)):*
> ```XML
> impl From<InvokeTransaction> for CentralInvokeTransactionV3 {
>     fn from(tx: InvokeTransaction) -> CentralInvokeTransactionV3 {
>         assert_matches!(tx.tx, starknet_api::transaction::InvokeTransaction::V3(_));
> ```

Change the input type to `InvokeTransactionV3`



_Code quote:_
```XML
assert_matches!(tx.tx, starknet_api::transaction::InvokeTransaction::V3(_));
```

**Review by Yael-Starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 4 unresolved discussions (waiting on @dafnamatsry, @DvirYo-starkware, and @graphite-app[bot])

___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 143 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OEwXh428AY5Losj4mW4:-OEx4WGGAD8Q96RbaBU4:byhcutv) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L143)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Change the input type to `InvokeTransactionV3`
</blockquote></details>

I thought about it but then I won't have the tx_hash, which is stored in the external object. 
I could send it separately but for other transactions I need more fields:
for declare_tx I need the tx_hash and the class_info
got deploy_account_tx I need the tx_hash and the contract address.

**Review by Yael-Starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 4 unresolved discussions (waiting on @dafnamatsry, @DvirYo-starkware, and @graphite-app[bot])

___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 134 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErtTiz-OgJ6LLHyHHD:-OEx5mSf9JkuihZKfrto:b-gflmvq) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L134)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Might be a bit risky to rely on complex objects that we are not managing a can potentially change.
</blockquote></details>

other objects of this struct are also from starknet_api.
The serialization test should catch it if any of them changes.

**Review by DvirYo-starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 9 unresolved discussions (waiting on @dafnamatsry, @graphite-app[bot], and @Yael-Starkware)

___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 5 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExGw098U3yU1LwxDPi:-OExGw098U3yU1LwxDPj:bwxhqj0) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L5)):*
> ```XML
> use indexmap::{indexmap, IndexMap};
> use serde::{Deserialize, Serialize};
> use serde_repr::*;
> ```

Why the use * thing?
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 104 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExHikS8qKUSqxdhDz7:-OExHikS8qKUSqxdhDz8:brkziei) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L104)):*
> ```XML
>     }
> }
> ```

As discussed, consider doing direct into instead of a middle type
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 110 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExI-q7AuKtjrJ2M46z:-OExI-q7AuKtjrJ2M47-:b-qif1q4) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L110)):*
> ```XML
> #[derive(Clone, Debug, Default, PartialEq, Deserialize_repr, Serialize_repr)]
> #[serde(into = "u8", try_from = "u8")]
> #[repr(u8)]
> ```

Why this?



_Code quote:_
```XML
#[repr(u8)]
```
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 112 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExH5ba2fnl6boBzRVe:-OExH5ba2fnl6boBzRVf:b-624b0j) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L112)):*
> ```XML
> #[repr(u8)]
> pub enum CentralDataAvailabilityMode {
>     #[default]
> ```

Why `default`?



_Code quote:_
```XML
#[default]
```
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs` line 95 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExIl2r3o4RBrNJ2uvc:-OExIl2r3o4RBrNJ2uvd:b-6n7lk2) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs#L95)):*
> ```XML
>             paymaster_data: Default::default(),
>             account_deployment_data: Default::default(),
>             nonce_data_availability_mode: Default::default(),
> ```

I really don't like the fact that we have so many `default` fields. We may miss here something that will not be tested.
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs` line 105 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExJCwO60Z9TGyWbMBB:-OExJCwO60Z9TGyWbMBC:bq2eawk) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs#L105)):*
> ```XML
> }
> 
> #[rstest]
> ```

The test structure looks very good

**Review by dafnamatsry — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 7 unresolved discussions (waiting on @graphite-app[bot] and @Yael-Starkware)

___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 134 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErtTiz-OgJ6LLHyHHD:-OExKV86BdfMYa4_5XWY:bzcwj56) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L134)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

other objects of this struct are also from starknet_api.
The serialization test should catch it if any of them changes.
</blockquote></details>

The other object are basic types (i.e. u64, etc.).
Having a `central` version will make it easier for us to adjust the serialisation in case the type changes.
O'm not insisting on that, but I think it should be considered.
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 143 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OEwXh428AY5Losj4mW4:-OExJCv30OfWQl8tEkmG:b5kq9yw) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L143)):*
<details><summary><i>Previously, Yael-Starkware (YaelD) wrote…</i></summary><blockquote>

I thought about it but then I won't have the tx_hash, which is stored in the external object. 
I could send it separately but for other transactions I need more fields:
for declare_tx I need the tx_hash and the class_info
got deploy_account_tx I need the tx_hash and the contract address.
</blockquote></details>

I see, ok then...

**Review by DvirYo-starkware — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 4 unresolved discussions (waiting on @graphite-app[bot] and @Yael-Starkware)

___
*[`crates/starknet_api/src/data_availability.rs` line 12 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExwnZeAyA8K5BW1bij:-OExwnZeAyA8K5BW1bik:bbjcpes) ([raw file](https://github.com/starkware-libs/sequencer/blob/92c3cf939c46986057cf8996a7943469c3e27405/crates/starknet_api/src/data_availability.rs#L12)):*
> ```XML
>     L2 = 1,
> }
> ```

Move this function to central_objects (you will not able to do the From trait there(

**Review by Yael-Starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 4 unresolved discussions (waiting on @dafnamatsry, @DvirYo-starkware, and @graphite-app[bot])

___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 134 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OErtTiz-OgJ6LLHyHHD:-OExqU7yBDZmtQAmsVwz:b-6xlnt1) ([raw file](https://github.com/starkware-libs/sequencer/blob/332d29e6c0be465dd2120cc4e24e64793d20b3a3/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L134)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

The other object are basic types (i.e. u64, etc.).
Having a `central` version will make it easier for us to adjust the serialisation in case the type changes.
O'm not insisting on that, but I think it should be considered.
</blockquote></details>

as discussed I added a todo to consider it later.
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 5 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExGw098U3yU1LwxDPi:-OExqJTnDVHgtcWQtG27:b-hfpv0l) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L5)):*
<details><summary><i>Previously, DvirYo-starkware wrote…</i></summary><blockquote>

Why the use * thing?
</blockquote></details>

removed
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 104 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExHikS8qKUSqxdhDz7:-OExqLJ60HkHG2Jfz2XA:bb7f2mf) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L104)):*
<details><summary><i>Previously, DvirYo-starkware wrote…</i></summary><blockquote>

As discussed, consider doing direct into instead of a middle type
</blockquote></details>

done
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 110 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExI-q7AuKtjrJ2M46z:-OExqMY-95ofKunWqUkD:b-jyon2k) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L110)):*
<details><summary><i>Previously, DvirYo-starkware wrote…</i></summary><blockquote>

Why this?
</blockquote></details>

code was deleted
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs` line 112 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExH5ba2fnl6boBzRVe:-OExqNu286myI6Cgn_RM:b-jyon2k) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects.rs#L112)):*
<details><summary><i>Previously, DvirYo-starkware wrote…</i></summary><blockquote>

Why `default`?
</blockquote></details>

code was deleted
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs` line 95 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExIl2r3o4RBrNJ2uvc:-OExqbXj9q7T9_c4QTFn:b38tdm2) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs#L95)):*
<details><summary><i>Previously, DvirYo-starkware wrote…</i></summary><blockquote>

I really don't like the fact that we have so many `default` fields. We may miss here something that will not be tested.
</blockquote></details>

I think it's fine since these are very simple fields. let me add a todo for later.
Changing the transaction requires generating a new object in python, so this can be a separate task
___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs` line 105 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExJCwO60Z9TGyWbMBB:-OExuqWZ1SksDBz6dJ74:bd5l6m4) ([raw file](https://github.com/starkware-libs/sequencer/blob/dfcae509b74a977f4231d122e164fc9dd4fa143d/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs#L105)):*
<details><summary><i>Previously, DvirYo-starkware wrote…</i></summary><blockquote>

The test structure looks very good
</blockquote></details>

thanks

**Review by DvirYo-starkware — APPROVED** [review]
> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 2 unresolved discussions (waiting on @dafnamatsry and @graphite-app[bot])

**Review by Yael-Starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 2 unresolved discussions (waiting on @dafnamatsry and @graphite-app[bot])

___
*[`crates/starknet_api/src/data_availability.rs` line 12 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OExwnZeAyA8K5BW1bij:-OEyEym33hOqOKu5GoBx:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/92c3cf939c46986057cf8996a7943469c3e27405/crates/starknet_api/src/data_availability.rs#L12)):*
<details><summary><i>Previously, DvirYo-starkware wrote…</i></summary><blockquote>

Move this function to central_objects (you will not able to do the From trait there(
</blockquote></details>

Done.

**Review by dafnamatsry — APPROVED** [review]
> Reviewed 2 of 5 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 2 unresolved discussions (waiting on @graphite-app[bot] and @Yael-Starkware)

___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs` line 39 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OF0fiF-3bXlLQaeLhar:-OF0fiF-3bXlLQaeLhas:b-gqlidb) ([raw file](https://github.com/starkware-libs/sequencer/blob/2fc1955f897d0d6b5bd53c5b89fe2388de863b2a/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs#L39)):*
> ```XML
> pub const CENTRAL_INVOKE_TX_JSON_PATH: &str = "central_invoke_tx.json";
> 
> fn central_state_diff() -> CentralStateDiff {
> ```

Make this return a Value as well (for consistency). Nonblocking.

**Review by Yael-Starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: all files reviewed, 2 unresolved discussions (waiting on @dafnamatsry and @graphite-app[bot])

___
*[`crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs` line 39 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/2861#-OF0fiF-3bXlLQaeLhar:-OF1M8VNDfPEKxZkt1_L:b-pxvlk3) ([raw file](https://github.com/starkware-libs/sequencer/blob/2fc1955f897d0d6b5bd53c5b89fe2388de863b2a/crates/sequencing/papyrus_consensus_orchestrator/src/cende/central_objects_test.rs#L39)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Make this return a Value as well (for consistency). Nonblocking.
</blockquote></details>

that is the TODO I added below.

**Review by Yael-Starkware — COMMENTED** [review]
> Dismissed @graphite-app[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2861)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Yael-Starkware)

**Pull request merged**